internal/controller/apiextensions/composite/composition_pt.go

```diff
@@ -273,6 +273,21 @@ func (c *PTComposer) Compose(ctx context.Context, xr *composite.Unstructured, re
 		o := []resource.ApplyOption{resource.MustBeControllableBy(xr.GetUID()), usage.RespectOwnerRefs()}
 		o = append(o, mergeOptions(filterPatches(t.Patches, patchTypesFromXR()...))...)
 		if err := c.client.Apply(ctx, cd, o...); err != nil {
+			if kerrors.IsInvalid(err) {
+				// We tried applying an invalid resource, we can't tell whether
+				// this means the resource will never be valid or it will if we
+				// run again the composition after some other resource is
+				// created or updated successfully. So, we emit a warning event
+				// and move on.
+				events = append(events, event.Warning(reasonCompose, errors.Wrap(err, errApplyComposed)))
+				// We unset the cd here so that we don't try to observe it
+				// later. This will also mean we report it as not ready and not
+				// synced. Resulting in the XR being reported as not ready nor
+				// synced too.
+				cds[i] = nil
```

Comment:
I actually realised this is probably not what we want. Originally setting an element of cds to nil meant the resource hadn't been created at all (because we couldn't dry-run it), so it made sense to skip observing it below, setting it to ready false as it didn't exist. But now it could mean we couldn't just update it, so we should still report it's readiness correctly if possible. But I'm struggling to find the right way to do it keeping the change as small as possible. Any thought @negz?

Reply:
One thing not clear to me is where in here we get the observed resources to feed into the patches, is the assumption that the Apply will also update the status instead of observing composed resources and feeding them to the composition logic as the function composer does? I can see only a [Get](https://github.com/crossplane/crossplane/blob/850a0d494c53393e9b94744597fe14e50081b599/internal/controller/apiextensions/composite/composition_pt.go#L481), but resources are not persisted anywhere, am I missing something?

Reply:
if we don't get observed resources it looks like that would be a regression w.r.t. 1.12 as [here](https://github.com/crossplane/crossplane/blob/13c2d1f42c61d332b065f3b20c1db6e10c773c37/internal/controller/apiextensions/composite/composition_pt.go#L220-L223) we would run patches to the environment without the resources status, while we did properly handle it in 1.12 [here](https://github.com/crossplane/crossplane/blob/5b491478367e20202ad85cb57ad2e94eebb0320a/internal/controller/apiextensions/composite/composition_ptf.go#L548).

Reply:
Makes sense to me, but I think we can start from this smaller change already solving the issues at hand and only later, if we feel it's worth the time, rework it more aggressively to fix https://github.com/crossplane/crossplane/issues/5388 too. If we agree this solution is good enough I can work on adding some e2e tests, so that we'll have some confidence refactoring this in the future.

Reply:
@phisco I'm all for a smaller change.

Weren't you saying though that there's still another issue in the `PTComposer`? If a composed resource exists, but we can't update it because it's now invalid, we won't be able to check its readiness and will thus report it's unready.

Is that right?

Reply:
Just chatted to @phisco on Zoom. There is still the other issue pointed out in the preceding comment, but we'll defer addressing that to another PR. Even without fixing that issue this PR still moves us forward to a better state.